contents/docs/libraries/remix.md

```diff
@@ -18,36 +18,143 @@ yarn add posthog-js
 
-```ts file=entry.client.tsx
-import { RemixBrowser } from "@remix-run/react";
-import { startTransition, StrictMode, useEffect } from "react";
-import { hydrateRoot } from "react-dom/client";
+
+```ts file=app/root.tsx
+import { Links, Meta, Outlet, Scripts, ScrollRestoration, useLoaderData } from "@remix-run/react";
+import type { LinksFunction } from "@remix-run/node";
+import { json } from "@remix-run/node";
+
+//... other imports, links, etc.
+
+export const loader = () => {
+  return json({
+    ENV: {
+      POSTHOG_API_KEY: process.env.POSTHOG_API_KEY,
+    },
+  });
+};
+
+export default function App() {
+  const { ENV } = useLoaderData<typeof loader>();
+  return (
+    <html lang="en">
+      <head>
+        <Meta />
+        <Links />
+      </head>
+      <body>
+        <Outlet />
+        <ScrollRestoration />
+        <script
+          dangerouslySetInnerHTML={{
+            __html: `window.ENV = ${JSON.stringify(ENV)}`,
+          }}
+        />
+        <Scripts />
+      </body>
+    </html>
+  );
+}
+```
+
+Next, create a new PostHog context in `app/contexts/posthog-context.tsx`. This is necessary because of a [missing export statement](https://github.com/PostHog/posthog-js/issues/908) in `posthog-js`'s `package.json`. 
+
+```ts file=app/contexts/posthog-context.tsx
 import posthog from "posthog-js";
+import React, { createContext, useContext, useRef } from "react";
 
-function PosthogInit() {
-  useEffect(() => {
-    posthog.init('<ph_project_api_key>', {
+type PosthogType = typeof posthog | undefined;
+
+const PosthogContext = createContext<PosthogType>(undefined);
+
+interface PosthogProviderProps {
+  children: React.ReactNode;
+}
+
+export function PosthogProvider({ children }: PosthogProviderProps) {
+  const posthogInstanceRef = useRef<PosthogType>(undefined);
+
+  // https://react.dev/reference/react/useRef#avoiding-recreating-the-ref-contents
+  // Note that in StrictMode, this will run twice.
+  function getPosthogInstance() {
+    if (posthogInstanceRef.current) return posthogInstanceRef.current;
+    if (!window.ENV.POSTHOG_API_KEY) return undefined;
+
+    posthogInstanceRef.current = posthog.init(window.ENV.POSTHOG_API_KEY, {
       api_host: '<ph_client_api_host>',
-      person_profiles: 'identified_only',
+      person_profiles: 'identified_only'
     });
-  }, []);
+    return posthogInstanceRef.current;
+  }
 
-  return null;
+  return (
+    <PosthogContext.Provider value={getPosthogInstance()}>
+      {children}
+    </PosthogContext.Provider>
+  );
 }
 
+export const usePosthog = () => useContext(PosthogContext);
+```
+
+Lastly, we need to add this context to your React tree. Go to `app/entry.client.tsx` and add the following: 
+
+```ts file=app/entry.client.tsx
```

Comment:
Ok, checking again, the `posthog-context` seems to work without hydration errors if you use it in the `app.root.tsx` instead of `entry.client`:

```
// app/root.tsx
import { PosthogProvider } from "./contexts/posthog-context";

export default function App() {
  const { ENV } = useLoaderData<typeof loader>();

  return (
    <html lang="en">
      <head>
        <Meta />
        <Links />
      </head>
      <body>
        <PosthogProvider>
          <Outlet />
        </PosthogProvider>
        <ScrollRestoration />
        <script
          dangerouslySetInnerHTML={{
            __html: `window.ENV = ${JSON.stringify(ENV)}`,
          }}
        />
        <Scripts />
      </body>
    </html>
  );
}
```

Reply:
Yeah, TBH I have no idea why you're experiencing these issues on Vercel and I'm not able to replicate the hydration errors running your repo locally (whether running `npm run dev` or `npm run build && npm run start` [prod]). Happy to take your lead here and update the PR to move it to root.tsx if you can confirm that it doesn't break on your vercel deployment! 

(note that remix-cool.vercel.app right now is running the dev version of your app for some reason, and doesn't show the hydration errors)

Reply:
None of these seem to work now 😢, let me try some more options.

Reply:
Couldn't figure it out yet, still getting a variety of hydration errors

Reply:
Could you publish your Remix app that works so I could test it out?